### DISCUSSION TITLE
Using OpenStreetMap data

### DISCUSSION THREAD

**Discussion by antbern** [karttapullautin/karttapullautin#144] [Q&A] [answered]
I am looking at the code that renders shapefiles to the generated map image and I'm having a hard time to get it to work.

First of all, OSM does not seem to allow exporting an area in shapefile format (anymore), only in `.pbf` format. After some research I found a shapefile download for the entire Sweden at [geofrabrik.de](https://download.geofabrik.de/europe/sweden.html) which I tried to use, but after processing all the shapefiles in there (it comes with multiple ones for different layers like water and road) I have no output on my smaller test-tile. I checked the `osm.txt` file and compared it with the content of the shapefiles and noticed that they do not agree (tags format differ), but after modifying and double-checking that the code actually drew something on the images I still see no output on my file.  I suspect that this might be an issue with coordinate systems and the fact that my tile has points in `EPSG:3006 SWEREF99 TM` and the shapefile data in some other, possibly lat/lng. 

I also tried to download a smaller extract from [extract.bbbike.org](https://extract.bbbike.org/?lang=en;sw_lng=13.657;sw_lat=59.8473;ne_lng=13.713;ne_lat=59.8687;format=shp.zip;city=21D004_66350_4250_25_tile;email=anton%40berneving.com;as=0.056146484375;pg=1;coords=;oi=1;layers=B00T;submit=;expire=) but also that without success (those files also have a different file structure inside the `zip` file, causing some issues with the automatic extraction). The tags in this data was also not aligning with the rules defined in `osm.txt`.

So, do you have any tips on how to get this working? Should we allow for coordinate transformations to happen within `kartapullautin`? Should we implement support for the native OSM [`pbf`](https://wiki.openstreetmap.org/wiki/PBF_Format) format (I have seen some existing crates out there), [OSM XML](https://wiki.openstreetmap.org/wiki/OSM_XML) (what you get from their [export page](https://www.openstreetmap.org/export#map=14/59.85497/13.68672)), or is there an easier way to get correct shapefiles from OSM data for use with `kartapullautin`?

Thanks! :100:

**Comment by dg0yt** [comment]
Are there are no rust binding for GDAL and PROJ? These are the standard open source libs for dealing with geospatial data and coordinates.

**Reply by antbern** [reply]
Indeed there seems to be relatively up-to-date crates with bindings for both GDAL and PROJ so that should not be an issue :rocket:

**Reply by rphlo** [reply]
https://github.com/georust/gdal

**Comment by stefankinell** [comment]
I might be wrong but if I remember correctly you should convert to the right coordinate system, and then add it to a zip file that is then placed in the In-directory. I found some old discussions at https://www.attackpoint.org/discussionthread.jsp/message_907729#message908370 that supported that memory.

On the other hand - do we know if this feature is really used? I played around a lot in the beginning with how to use kartapullautin and ended up compiling the map in OOM or OCAD instead of making ready map images from KP. It just was easier and allowed for more control of what ended up on the map.

**Comment by rphlo** [comment, marked as answer]
My understanding is that at the moment you can download the files in .osm format then convert them to shapefiles in the same projection system of your .laz with, for example, QGIS...
The process is explained here: https://mapant.fr/blog/karttapullautin-pour-les-nuls (Translation tool may help)

Supporting .pbf format out of the box could be a nice feature if you feel like implementing it.

**Reply by antbern** [reply]
Thanks for the link, I will try it out! But it does make sense that one need to make sure the coordinates are in the same system... 

As @stefankinell pointed out above, perhaps it is not such a used feature given that it might be comubersome to generate the map when there are easier ways like using OOM or Qgis. Or perhaps it's the other way around, more people would use it if it was easier, like if extracts could be automatically downloaded etc. I'll see if I can get it working and if it makes sense to improve format support then that might be a good feature improvement 💯

**Reply by dg0yt** [reply]
> QGIS...

... will call GDAL and PROJ. OOM uses GDAL and PROJ, too.

**Reply by antbern** [reply]
> > QGIS...
> 
> ... will call GDAL and PROJ. OOM uses GDAL and PROJ, too.
> 

Well it seems GDAL and PROJ are the defacto libraries for doing spatial stuff :rocket: 🌍 Probably makes sense to use them :thumbsup:

**Reply by rphlo** [reply]
https://github.com/NicoRio42/cassini/blob/main/src/vectors.rs Nicolas Rio project is doing osm import using ogr2ogr call with hardcoded projection for France.
His project also include OSM data download https://github.com/NicoRio42/cassini/blob/main/src/download.rs

**Reply by dg0yt** [reply]
FTR: ogr2ogr is a binary from GDAL.
https://gdal.org/en/latest/programs/ogr2ogr.html#ogr2ogr

**Comment by rphlo** [comment]
I updated the README.txt on how to deal with .osm files.
